### PR TITLE
feat: DAH-2876 useRedirect for authenticated pages

### PR DESCRIPTION
**Pull request by cade-exygy**
## Description

If a user tries to go to the my-account, account-settings, or my-applications and is not signed in: redirect to sign in page, show banner. Once the login happens, the user is redirected to the page they were trying to access prior.

## Jira ticket

https://sfgovdt.jira.com/browse/DAH-2876

### Version Control

- [x] branch name begins with `angular` if it contains updates to Angular code
- [x] branch name contains the Jira ticket number
- [x] PR name follows `type: TICKET-NUMBER Description` format, e.g. `feat: DAH-123 New Feature`

### Code quality

- [x] [the set of changes is small](https://google.github.io/eng-practices/review/developer/small-cls.html#what-is-small)
- [ ] all automated code checks pass (linting, tests, coverage, etc.)
- [x] code irrelevant to the ticket is not modified e.g. changing indentation due to automated formatting
- [x] if the code changes the UI, it matches the UI design exactly
- [x] if the changes include human translations, follow the [human translations process](https://sfgovdt.jira.com/l/cp/XS1KpvE4)

### Review instructions

- [x] instructions specify which environment(s) it applies to
- [x] instructions work for PA testers
- [x] instructions have already been performed at least once

### Request review

- [x] PR has `needs review` label
- [x] Use `Housing Eng` group to automatically assign reviewers, and/or assign specific engineers
- [x] If time sensitive, notify engineers in Slack

## Review Instructions
- On the review app:
- While signed out, manually enter the restricted url: /my-account?react=true
- It should try to load the page, but then redirect back to /sign-in
- You should see the banner
- Sign In to an account: test@test.com / test1234
- Once you are signed in, you should be redirected
- Try this with /my-account, /account-settings, and /my-applications

Test edge cases like signing out, triggering other banners, etc

### CODE REVIEW

File: app/javascript/hooks/useRedirect.tsx [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+    sessionStorage.removeItem("redirect")
+    if (redirect) {
+      setRedirect(redirect)
+    } else {
+      setRedirect(null)
+      clearSiteAlertMessage("secondary")
+    }
+  }, [])
+
+  return { redirect, handleRedirect }
```
Comment (jimlin-sfgov) [inline app/javascript/hooks/useRedirect.tsx:31]: question: You're returning `redirect` here, but I don't see it getting used anywhere. Is there a future use for it?

File: app/javascript/pages/account/account-settings.tsx [hunk 2]
```diff
@@ -454,9 +455,10 @@ const AccountSettings = ({ profile }: { profile: User }) => {
 
 const AccountSettingsPage = () => {
   const { profile, loading, initialStateLoaded } = React.useContext(UserContext)
+  const { handleRedirect } = useRedirect()
 
   if (!profile && !loading && initialStateLoaded) {
-    // TODO: Redirect to React sign in page and show a message that user needs to sign in
+    handleRedirect("settings")
```
Comment (jimlin-sfgov) [inline app/javascript/pages/account/account-settings.tsx:461]: issue: not sure what exactly is happening, but this block of code is running twice, and causing the site alert message to be removed from storage, which causes the alert banner to never show up.

**Review by jimlin-sfgov — CHANGES_REQUESTED** [review]

File: app/javascript/hooks/useRedirect.tsx [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+
+const handleRedirect = (path: string) => {
+  const isRedirectSet = sessionStorage.getItem("redirect")
+  if (!isRedirectSet) {
+    sessionStorage.setItem("redirect", path)
+
+    setSiteAlertMessage(t("signIn.loginRequired"), "secondary")
+  } else {
+    sessionStorage.removeItem("redirect")
+    clearSiteAlertMessage("secondary")
```
Comment (jimlin-sfgov) [inline app/javascript/hooks/useRedirect.tsx:13]: suggestion: create a `useRedirect.test.tsx` file, if adding other tests does not reach the minimum coverage threshold.

File: app/javascript/util/routeUtil.ts [hunk 4]
```diff
@@ -75,3 +75,18 @@ export const getEligibilityEstimatorLink = localizedPathGetter("/eligibility-est
...
+  home: getHomepagePath(),
+}
+
+const getRedirectUrl = (key: string): string => {
+  return SignInRedirects[key] || SignInRedirects.home
+}
+
+export const getSignInRedirectUrl = (redirect: string) => {
+  return getRedirectUrl(redirect || "account")
+}
```
Comment (jimlin-sfgov) [inline app/javascript/util/routeUtil.ts:92]: issue: Add tests in for new functions in `routeUtil.test.ts`

File: app/javascript/authentication/SignInForm.tsx [hunk 5]
```diff
@@ -130,7 +133,7 @@ const SignInForm = () => {
 
     signIn(email, password)
       .then(() => {
-        window.location.href = getMyAccountPath()
+        window.location.href = getSignInRedirectUrl(redirect)
```
Comment (jimlin-sfgov) [inline app/javascript/authentication/SignInForm.tsx:136]: suggestion: add tests to the `SignIn.test.tsx` file, if adding other tests does not reach the minimum coverage threshold.

**Review by jimlin-sfgov — COMMENTED** [review]
these are just about coverage

**Review by tallulahkay — COMMENTED** [review]
I tried to review this, but I'm still not seeing the banner pop up on the sign in page when i type in my-account?react=true

**Comment by jimlin-sfgov** [comment]
I don't see the banner when following the review instructions in the review app or locally. @cade-exygy did you run through your review instructions before requesting review?

**Comment by cade-exygy** [comment]
> I don't see the banner when following the review instructions in the review app or locally. @cade-exygy did you run through your review instructions before requesting review?

@tallulahkay @jimlin-sfgov Added back the banner. FYI, the main changes here is the redirect logic, so looking for a review on that behavior.

**Review by tallulahkay — APPROVED** [review]
Nice!

File: app/javascript/__tests__/pages/account/account-settings.test.tsx [hunk 6]
```diff
@@ -70,20 +84,6 @@ describe("<AccountSettingsPage />", () => {
       expect(title).not.toBeNull()
     })
 
-    it("resize events", () => {
```
Comment (tallulahkay) [inline app/javascript/__tests__/pages/account/account-settings.test.tsx:73]: did you mean to delete this test?

File: app/javascript/authentication/withAuthentication.tsx [hunk 7]
```diff
@@ -0,0 +1,55 @@
...
+  { redirectPath }: WithAuthenticationProps = {}
+) => {
+  const WithAuthenticationComponent = (props: P) => {
+    const { profile, loading } = React.useContext(UserContext)
+
+    React.useEffect(() => {
+      if (!isTokenValid()) {
+        const redirectParam = redirectPath ? `?redirect=${redirectPath}` : ""
+        const language = getCurrentLanguage()
+        const signInPath = getLocalizedPath("/sign-in", language, redirectParam)
```
Comment (jimlin-sfgov) [inline app/javascript/authentication/withAuthentication.tsx:33]: thought: Storing redirect information in the URL is a departure from the previous iterations of this PR, and from the Angular version of the website. I'd love to get a breakdown of this decision. CC @chadbrokaw since I believe he is also quite involved with the authentication work.

**Review by jimlin-sfgov — APPROVED** [review]
LGTM

just needs coverage and conflict resolution!

File: app/javascript/authentication/withAuthentication.tsx (hunk 7 above)
Comment (cade-exygy) [reply app/javascript/authentication/withAuthentication.tsx:33]: Yeah for sure! In short, handling the redirect with url params allows us to avoid using session storage to keep track of the redirect state across page reloads. I feel like this is a more "React like" way of handling this, since the SignIn component now fully controls when to show the banner and adjust the redirect after submit, instead of relying on a session storage value being set (and then cleared appropriately). Since the banner is toggled only on the presence of the redirect flag, it is more clear why and when that happens imo. The withAuthentication HOC also provides us a way to handle adding this param for any page that we want to be gated behind an authentication state.
